File: risk-engine.js
/**
 * Client-side demo scenario definitions for the Live AI Drive page.
 * Each scenario sends a signal payload to the AI service's /predict-risk
 * endpoint (with a Node-side fallback — see backend/services/riskEngine.js)
 * and the AI service returns an explainable score. This file just defines
 * what each demo button simulates and drives the voice co-driver alerts.
 */

const SCENARIOS = {
  NORMAL: {
    label: "Normal Driving",
    icon: "bi-check-circle",
    signals: { speed: 55, speedLimit: 60, driverAttention: 95, weather: "CLEAR", roadCondition: "GOOD", trafficDensity: "LOW" },
    voice: { en: "All clear. Conditions are normal.", ta: "எல்லாம் சரி. நிலைமை இயல்பாக உள்ளது." },
  },
  OVERSPEEDING: {
    label: "Overspeeding",
    icon: "bi-speedometer",
    signals: { speed: 105, speedLimit: 80, driverAttention: 85, weather: "CLEAR", roadCondition: "GOOD", trafficDensity: "MODERATE" },
    voice: { en: "Warning. Reduce speed.", ta: "எச்சரிக்கை. வேகத்தைக் குறைக்கவும்." },
  },
  DROWSINESS: {
    label: "Drowsiness",
    icon: "bi-emoji-dizzy",
    signals: { speed: 70, speedLimit: 80, driverAttention: 30, drowsiness: true, weather: "CLEAR", roadCondition: "GOOD" },
    voice: { en: "Driver attention is low. Please take a break.", ta: "ஓட்டுநரின் கவனம் குறைவாக உள்ளது." },
  },
  DISTRACTION: {
    label: "Distraction",
    icon: "bi-eye-slash",
    signals: { speed: 65, speedLimit: 80, driverAttention: 40, distraction: true, weather: "CLEAR", roadCondition: "GOOD" },
    voice: { en: "Focus on the road.", ta: "சாலையில் கவனம் செலுத்துங்கள்." },
  },
  MOBILE_USAGE: {
    label: "Mobile Phone Usage",
    icon: "bi-phone",
    signals: { speed: 60, speedLimit: 80, driverAttention: 35, mobileUsage: true, weather: "CLEAR", roadCondition: "GOOD" },
    voice: { en: "Mobile phone usage detected. Please focus on driving.", ta: "மொபைல் பயன்பாடு கண்டறியப்பட்டது." },
  },
  POTHOLE: {
    label: "Pothole",
    icon: "bi-exclamation-diamond",
    signals: { speed: 60, speedLimit: 80, driverAttention: 90, roadCondition: "DAMAGED", weather: "CLEAR" },
    voice: { en: "Pothole detected ahead.", ta: "முன்னால் பள்ளம் உள்ளது." },
  },
  UNSAFE_OVERTAKING: {
    label: "Unsafe Overtaking",
    icon: "bi-arrow-left-right",
    signals: { speed: 85, speedLimit: 80, driverAttention: 80, followingDistanceMeters: 6, trafficDensity: "HIGH", weather: "CLEAR" },
    voice: { en: "Unsafe following distance.", ta: "பாதுகாப்பற்ற தொடர் தூரம்." },
  },
  WRONG_SIDE: {
    label: "Wrong Side Driving",
    icon: "bi-sign-turn-left",
    signals: { speed: 50, speedLimit: 60, driverAttention: 80, wrongSideDriving: true, weather: "CLEAR" },
    voice: { en: "Warning. You are on the wrong side of the road.", ta: "எச்சரிக்கை. தவறான பக்கத்தில் உள்ளீர்கள்." },
  },
  HEAVY_RAIN: {
    label: "Heavy Rain",
    icon: "bi-cloud-rain-heavy",
    signals: { speed: 70, speedLimit: 80, driverAttention: 85, weather: "HEAVY_RAIN", roadCondition: "WET", trafficDensity: "MODERATE" },
    voice: { en: "Heavy rain detected. Recommended speed reduced.", ta: "கடும் மழை கண்டறியப்பட்டது." },
  },
  LANE_DEPARTURE: {
    label: "Lane Departure",
    icon: "bi-sign-merge-left",
    signals: { speed: 75, speedLimit: 80, driverAttention: 60, laneDeparture: true, weather: "CLEAR" },
    voice: { en: "Lane departure detected. Please steer back.", ta: "பாதை விலகல் கண்டறியப்பட்டது." },
  },
  SUDDEN_BRAKING: {
    label: "Sudden Braking",
    icon: "bi-slash-circle",
    signals: { speed: 65, speedLimit: 80, driverAttention: 80, suddenBraking: true, trafficDensity: "HIGH", weather: "CLEAR" },
    voice: { en: "Sudden braking event recorded.", ta: "திடீர் பிரேக் நிகழ்வு பதிவு செய்யப்பட்டது." },
  },
  COLLISION_RISK: {
    label: "Collision Risk",
    icon: "bi-exclamation-octagon",
    signals: { speed: 90, speedLimit: 80, driverAttention: 50, followingDistanceMeters: 5, distraction: true, trafficDensity: "HIGH", weather: "CLEAR" },
    voice: { en: "Collision risk detected. Reduce speed immediately.", ta: "மோதல் அபாயம் கண்டறியப்பட்டுள்ளது. உடனடியாக வேகத்தைக் குறைக்கவும்." },
  },
  ACCIDENT: {
    label: "Accident",
    icon: "bi-车",
    signals: { speed: 40, speedLimit: 80, driverAttention: 20, suddenBraking: true, wrongSideDriving: false, weather: "CLEAR" },
    voice: { en: "Emergency event detected. Checking driver response.", ta: "அவசர நிகழ்வு கண்டறியப்பட்டது." },
  },
};

module.exports = typeof module !== "undefined" ? { SCENARIOS } : undefined;

File: ai-service/services/risk_engine.py
"""
Explainable, configurable risk engine.

Mirrors backend/services/riskEngine.js so both Node and Python sides agree
on how a risk score is derived. This is the DEMO / rule-based engine that
always works without any trained model. Real AI mode (see services/real_ai.py
stub) would eventually feed its outputs into the same explainable structure.
"""
from utils.config import risk_level_from_score

WEIGHTS = {
    "overspeeding_per_kmh_over": 1.2,
    "overspeeding_cap": 30,
    "low_attention": 25,
    "drowsiness": 30,
    "distraction": 20,
    "mobile_usage": 25,
    "heavy_rain": 20,
    "fog": 20,
    "poor_road_condition": 15,
    "high_traffic_density": 10,
    "low_following_distance": 15,
    "lane_departure": 15,
    "wrong_side_driving": 35,
    "sudden_braking": 10,
    "night_low_visibility": 10,
}


def clamp(value, lo, hi):
    return max(lo, min(hi, value))


def calculate_risk(req) -> dict:
    breakdown = {}
    reasons = []
    score = 0.0

    over = max(0, (req.speed or 0) - (req.speedLimit or 0))
    if over > 0:
        pts = clamp(over * WEIGHTS["overspeeding_per_kmh_over"], 0, WEIGHTS["overspeeding_cap"])
        score += pts
        breakdown["overspeeding"] = round(pts)
        reasons.append(f"Overspeeding by {over:.0f} km/h")

    if req.driverAttention is not None and req.driverAttention < 50:
        score += WEIGHTS["low_attention"]
        breakdown["lowAttention"] = WEIGHTS["low_attention"]
        reasons.append("Low driver attention")

    if req.drowsiness:
        score += WEIGHTS["drowsiness"]
        breakdown["drowsiness"] = WEIGHTS["drowsiness"]
        reasons.append("Driver drowsiness detected")

    if req.distraction:
        score += WEIGHTS["distraction"]
        breakdown["distraction"] = WEIGHTS["distraction"]
        reasons.append("Driver distraction")

    if req.mobileUsage:
        score += WEIGHTS["mobile_usage"]
        breakdown["mobileUsage"] = WEIGHTS["mobile_usage"]
        reasons.append("Mobile phone usage while driving")

    if req.weather in ("HEAVY_RAIN", "RAIN"):
        score += WEIGHTS["heavy_rain"]
        breakdown["weather"] = WEIGHTS["heavy_rain"]
        reasons.append("Heavy rain")
    elif req.weather == "FOG":
        score += WEIGHTS["fog"]
        breakdown["weather"] = WEIGHTS["fog"]
        reasons.append("Fog / low visibility weather")

    if req.roadCondition in ("DAMAGED", "UNDER_CONSTRUCTION"):
        score += WEIGHTS["poor_road_condition"]
        breakdown["roadCondition"] = WEIGHTS["poor_road_condition"]
        reasons.append("Poor road condition")

    if req.trafficDensity == "HIGH":
        score += WEIGHTS["high_traffic_density"]
        breakdown["traffic"] = WEIGHTS["high_traffic_density"]
        reasons.append("High traffic density")

    if req.followingDistanceMeters is not None and req.followingDistanceMeters < 15:
        score += WEIGHTS["low_following_distance"]
        breakdown["followingDistance"] = WEIGHTS["low_following_distance"]
        reasons.append("Low following distance")

    if req.laneDeparture:
        score += WEIGHTS["lane_departure"]
        breakdown["laneDeparture"] = WEIGHTS["lane_departure"]
        reasons.append("Lane departure detected")

    if req.wrongSideDriving:
        score += WEIGHTS["wrong_side_driving"]
        breakdown["wrongSideDriving"] = WEIGHTS["wrong_side_driving"]
        reasons.append("Wrong-side driving detected")

    if req.suddenBraking:
        score += WEIGHTS["sudden_braking"]
        breakdown["suddenBraking"] = WEIGHTS["sudden_braking"]
        reasons.append("Sudden braking event")

    if req.timeOfDay == "NIGHT" and req.visibilityMeters is not None and req.visibilityMeters < 100:
        score += WEIGHTS["night_low_visibility"]
        breakdown["lowVisibility"] = WEIGHTS["night_low_visibility"]
        reasons.append("Low visibility at night")

    risk_score = round(clamp(score, 0, 100))
    risk_level = risk_level_from_score(risk_score)

    if not reasons:
        reasons.append("No significant risk factors detected")

    confidence = round(clamp(0.6 + len(breakdown) * 0.05, 0.6, 0.97), 2)

    recommendations = build_recommendations(risk_level, reasons)
    recommended_speed = build_recommended_speed(req, risk_level)

    return {
        "status": "success",
        "riskScore": risk_score,
        "riskLevel": risk_level,
        "confidence": confidence,
        "reasons": reasons,
        "breakdown": breakdown,
        "recommendations": recommendations,
        "recommendedSpeed": recommended_speed,
        "isDemoMode": True,
    }


def build_recommendations(risk_level, reasons):
    if risk_level == "CRITICAL":
        return [
            "Reduce speed immediately",
            "Maintain a safe following distance",
            "Focus fully on the road",
        ]
    if risk_level == "HIGH":
        return ["Slow down and stay alert", "Conditions indicate elevated risk: " + ", ".join(reasons)]
    if risk_level == "MODERATE":
        return ["Drive cautiously", "Monitor changing conditions"]
    return ["Conditions are favorable", "Continue safe driving practices"]


def build_recommended_speed(req, risk_level):
    limit = req.speedLimit or 60
    if risk_level == "CRITICAL":
        return round(limit * 0.6)
    if risk_level == "HIGH":
        return round(limit * 0.75)
    if risk_level == "MODERATE":
        return round(limit * 0.9)
    return round(limit)
